File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/with morfo library/chocolate_feast.ts
// https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

import { read, readLn } from 'morfo';

for (let i = 0, n = read(Number); i < n; i++) {
  const [money, cost, wrappersNeeded] = readLn(Number);
  console.log(howManyChocolatesCanBeEaten(money, cost, wrappersNeeded));
}

// T: O(log(money / cost))
// S: O(1) extra space
function howManyChocolatesCanBeEaten(
  money: number,
  cost: number,
  wrappersNeeded: number,
): number {
  let chocolates: number = Math.trunc(money / cost);

  for (let wrappers = chocolates; wrappers >= wrappersNeeded; ) {
    const freeChocolates = Math.trunc(wrappers / wrappersNeeded);
    wrappers = (wrappers % wrappersNeeded) + freeChocolates;
    chocolates += freeChocolates;
  }

  return chocolates;
}

File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/with morfo library/chocolate_feast.rs
// https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

use morfo::{read, read_ln};

fn main() {
    let n: usize = read();
    for _ in 0..n {
        let vals: Vec<i32> = read_ln();
        let (money, cost, wrappers_needed) = (vals[0], vals[1], vals[2]);
        println!(
            "{}",
            how_many_chocolates_can_be_eaten(money, cost, wrappers_needed)
        );
    }
}

// T: O(log(money / cost))
// S: O(1) extra space
fn how_many_chocolates_can_be_eaten(money: i32, cost: i32, wrappers_needed: i32) -> i32 {
    let mut chocolates: i32 = money / cost;
    let mut wrappers: i32 = chocolates;

    while wrappers >= wrappers_needed {
        let free_chocolates: i32 = wrappers / wrappers_needed;
        wrappers = wrappers % wrappers_needed + free_chocolates;
        chocolates += free_chocolates;
    }

    chocolates
}
